ibm_cloud_sdk_core/token_manager.py

```diff
@@ -0,0 +1,246 @@
+# coding: utf-8
+
+# Copyright 2020 IBM All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-instance-attributes
+
+
+from abc import ABC
+from threading import Lock
+from typing import Optional
+import time
+import requests
+
+from .api_exception import ApiException
+
+
+class TokenManager(ABC):
+    """An abstract class to contain functionality for parsing, storing, and requesting tokens.
+
+    get_token will retrieve a new token from the url in case the that there is no existing token,
+    or the previous token has expired. Child classes will implement request_token, which will do
+    the actual acquisition of a new token.
+
+    Args:
+        url: The url to request tokens from.
+
+    Keyword Args:
+        disable_ssl_verification: A flag that indicates whether verification of
+            the server's SSL certificate should be disabled or not. Defaults to False.
+        token_name: The key that maps to the token in the dictionary returned from request_token. Defaults to None.
+
+    Attributes:
+        url (str): The url to request tokens from.
+        disable_ssl_verification (bool): A flag that indicates whether verification of
+        the server's SSL certificate should be disabled or not.
+        token_name (str): The key used of the token in the dict returned from request_token.
+        token_info (dict): The most token_response from request_token.
+        expire_time (int): The time in epoch seconds when the current token within token_info will expire.
+        refresh_time (int): The time in epoch seconds when the current token within token_info should be refreshed.
+        request_time (int): The time the last outstanding token request was issued
+        lock (Lock): Lock variable to serialize access to refresh/request times
```

Comment:
Is there a reason we expose `lock` as an attribute?  Seems like this should be just internal state.

Reply:
It was already exposed in [JWTTokenManager](https://github.com/IBM/python-sdk-core/blob/master/ibm_cloud_sdk_core/jwt_token_manager.py#L50). That allows "external" code to utilize the lock for synchronization. That said I'm okay with moving it to internal. Let me know if you'd like me to make it "internal use".

Reply:
Good point.  I suppose it would technically be a breaking change to make it internal, and while I think the chance of actually breaking anyone is low, there's more risk than value in making the change so we should leave it as is.